### PR TITLE
C++: Fix unnecessary evaluation of debug strings

### PR DESCRIPTION
**Pull request by MathiasVP**
For some reason, the evaluator was sometimes evaluating the debug `toString` we added in https://github.com/github/codeql/pull/15088 even though they were never needed (which can be a huge performance problem, and can even cause the analysis to go into an infinite loop).

This PR fixes the problem by replacing the use of abstract classes and predicates (for controlling whether the debug or non-debug string should be produced) with a parameterized module. I've confirmed that this causes the queries to no longer produce the debug strings when debugging isn't enabled.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DebugPrinting.qll [hunk 1]
```diff
@@ -1,9 +1,21 @@
 /**
- * This file activates debugging mode for dataflow node printing.
+ * This file contains module that implements the _debug_ version of
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DebugPrinting.qll:2]:
```suggestion
 * This file contains a module that implements the _debug_ version of
```

**Review by jketema — COMMENTED** [review]
I see a signature, but no parameterized module. Am I missing something?

**Comment by MathiasVP** [comment]
> I see a signature, but no parameterized module. Am I missing something?

Oh, sorry. Yeah, I see that I wrote "parameterized module", but I really meant "good old boring module without any parameters". So the signature isn't really needed. I mostly used it to not duplicate the QLDoc in the two files that implement the signature (i.e., `NormalNode0ToString.qll` and `DebugPrinting.qll`).

**Comment by jketema** [comment]
> > I see a signature, but no parameterized module. Am I missing something?
> 
> Oh, sorry. Yeah, I see that I wrote "parameterized module", but I really meant "good old boring module without any parameters". So the signature isn't really needed. I mostly used it to not duplicate the QLDoc in the two files that implement the signature (i.e., `NormalNode0ToString.qll` and `DebugPrinting.qll`).

Wouldn't an abstract class be better?

**Comment by MathiasVP** [comment]
It would. That's basically what we had in https://github.com/github/codeql/pull/15088. But I couldn't make the evaluator not evaluate the `DebugPrinting` version with an abstract class 😭

**Comment by jketema** [comment]
> It would. That's basically what we had in #15088. But I couldn't make the evaluator not evaluate the `DebugPrinting` version with an abstract class 😭

So my intuition was - but this may be wrong - that #15088 was doing something more complex with `if`s that checked whether we were in debugging mode, and that this might be causing the unwanted evaluation. That is going now though, so I would have expected that using a abstract class instead of a signature in immaterial as long as we do not import the debug module anywhere.

**Comment by MathiasVP** [comment]
> So my intuition was - but this may be wrong - that #15088 was doing something more complex with `if`s that checked whether we were in debugging mode, and that this might be causing the unwanted evaluation. That is going now though, so I would have expected that using a abstract class instead of a signature in immaterial as long as we do not import the debug module anywhere.

Yeah, maybe you're right. I'll see what happens with this design if I replace the module signature with an abstract class and the modules with a concrete class 🤞

**Comment by MathiasVP** [comment]
@jketema you were correct 🎉. This design works well with an abstract class!

**Review by jketema — APPROVED** [review]
LGTM if DCA is happy.

**Comment by MathiasVP** [comment]
DCA looks uneventful 🎉

**Pull request merged**